### PR TITLE
Feature/excel-download

### PR DESCRIPTION
**Pull request by Mark-de-Haan**

### CODE REVIEW

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+						}
+					});
+
+			if (dataRequest.getColNames() == ColNames.ATTRIBUTE_LABELS)
+			{
+				csvWriter.writeAttributes(attributes);
+			}
+			else if (dataRequest.getColNames() == ColNames.ATTRIBUTE_NAMES)
+			{
+				csvWriter.writeAttributeNames(Iterables.transform(attributes, new Function<AttributeMetaData, String>()
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java]: csvWriter.writeAttributeNames(Iterables.transform(attributes, AttributeMetaData::getName));

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java [hunk 2]
```diff
@@ -412,18 +410,34 @@ public void updateMenuOnDeleteEntity(@PathVariable("entityName") String entityNa
...
+			String fileName = dataRequest.getEntityName() + '_'
+					+ new SimpleDateFormat("yyyy-MM-dd hh:mm:ss").format(new Date()) + ".csv";
+			response.addHeader("Content-Disposition", "attachment; filename=" + fileName);
+
+			ServletOutputStream outputStream = response.getOutputStream();
+			download.writeToCsv(dataRequest, outputStream, ',');
+		}
+		else if (dataRequest.getDownloadType() == DownloadType.DOWNLOAD_TYPE_XLSX)
+		{
+			response.setContentType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java]: Spring has content type negotiation and we probably can use that here. The idea is that you let the controller return a model object that contains all the information that should be rendered. Then Spring determines which serializer to use based on the requested content type in the header or in the extension of the request URL.

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+		Iterable<AttributeMetaData> attributes = dataService.getEntityMetaData(entityName).getAtomicAttributes();
+		List<String> attributeNames = new ArrayList<String>();
+
+		attributes.forEach(attribute -> attributeNames.add(attribute.getName()));
+
+		try
+		{
+			Writable writable = writableFactory.createWritable(entityName, attributeNames);
+			writable.add(repo);
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java]: The CSV exporter filters using query, the Excel exporter writes the entire repo.

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java [hunk 4]
```diff
@@ -456,55 +472,6 @@ public void exportToGalaxy(@Valid @RequestBody GalaxyDataExportRequest galaxyDat
 		model.addAttribute(ATTR_GALAXY_API_KEY, galaxyApiKey);
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java]: exportToGalaxy remains unchanged?

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java [hunk 5]
```diff
@@ -412,18 +410,34 @@ public void updateMenuOnDeleteEntity(@PathVariable("entityName") String entityNa
...
+			response.setContentType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
+			String fileName = dataRequest.getEntityName() + '_'
+					+ new SimpleDateFormat("yyyy-MM-dd hh:mm:ss").format(new Date()) + ".xlsx";
+			response.addHeader("Content-Disposition", "attachment; filename=" + fileName);
 
-		response.setContentType("text/csv");
-		response.addHeader("Content-Disposition", "attachment; filename=" + fileName);
-		writeDataRequestCsv(dataRequest, response.getOutputStream(), ',');
+			ServletOutputStream outputStream = response.getOutputStream();
+			download.writeToExcel(dataRequest, outputStream, '\t');
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java]: tab-separated excel!

File: molgenis-dataexplorer/src/main/resources/js/dataexplorer-data.js [hunk 6]
```diff
@@ -92,7 +92,8 @@
 			query : {
 				rules : [entityQuery.q]
 			},
-			colNames : $('input[name=colNames]:checked').val()
+			colNames : $('input[name=colNames]:checked').val(),
+			downloadType : $('input[name=downloadTypes]:checked').val()
 		};
```
Comment (fdlk) [inline molgenis-dataexplorer/src/main/resources/js/dataexplorer-data.js]: Does this make sense in the case of a galaxy export?
And in the case of excel, what is supposed to happen when the attribute labels are chosen as column names?

File: molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java (hunk 3 above)
Comment (Mark-de-Haan) [inline molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/download/DataExplorerDownloadHandler.java]: Doesn't this happen in the javascript?

File: molgenis-dataexplorer/src/main/resources/js/dataexplorer-data.js (hunk 6 above)
Comment (Mark-de-Haan) [inline molgenis-dataexplorer/src/main/resources/js/dataexplorer-data.js]: Galaxy export is a different button, but yes it is always csv. For excel the same thing happens as with csv, if someone chooses labels or names the column headers change

**Pull request merged**